File: api/app/domain/flares.py
"""Grouping symptoms into flares, and finding the meals that preceded them."""

from dataclasses import dataclass, field
from datetime import datetime, timedelta

from app.domain.rows import MealRow, SymptomRow
from app.domain.severity import intensity_from_severity
from app.domain.time import as_utc

# Symptoms logged within this long of the flare's start are the same flare
FLARE_CLUSTER_MINUTES = 30


def intensity_of(symptom: SymptomRow) -> int:
    if symptom.intensity is not None:
        return symptom.intensity
    return intensity_from_severity(symptom.severity)


@dataclass(eq=False)  # identity hashing: flares are collected in sets
class Flare:
    start: datetime
    end: datetime
    symptoms: list[SymptomRow] = field(default_factory=list)
    names: list[str] = field(default_factory=list)
    max_intensity: int = 0


def cluster_flares(symptoms: list[SymptomRow]) -> list[Flare]:
    """One flare per burst of symptoms.

    The window is measured from the flare's start, not the previous symptom,
    so a long trickle of entries can't chain into one endless flare.
    """
    window = timedelta(minutes=FLARE_CLUSTER_MINUTES)
    flares: list[Flare] = []
    for symptom in sorted(symptoms, key=lambda s: as_utc(s.timestamp)):
        at = as_utc(symptom.timestamp)
        current = flares[-1] if flares else None
        if current is not None and at - current.start <= window:
            current.symptoms.append(symptom)
            current.end = at
            if symptom.name not in current.names:
                current.names.append(symptom.name)
            current.max_intensity = max(current.max_intensity, intensity_of(symptom))
        else:
            flares.append(
                Flare(
                    start=at,
                    end=at,
                    symptoms=[symptom],
                    names=[symptom.name],
                    max_intensity=intensity_of(symptom),
                )
            )
    return flares


def hours_between(earlier: datetime, later: datetime) -> float:
    return (as_utc(later) - as_utc(earlier)).total_seconds() / 3600


def meals_before(meals: list[MealRow], at: datetime, window_hours: float) -> list[MealRow]:
    """Meals eaten in the window before `at`, most recent first.

    Strictly before: a meal logged at the same instant as the symptom cannot
    have caused it.
    """
    within = [m for m in meals if 0 < hours_between(m.timestamp, at) <= window_hours]
    return sorted(within, key=lambda m: as_utc(m.timestamp), reverse=True)


def onset_hours(meals: list[MealRow], at: datetime, window_hours: float) -> float | None:
    """Hours from the most recent meal in the window, or None if there was none."""
    latest = meals_before(meals, at, window_hours)
    return hours_between(latest[0].timestamp, at) if latest else None

File: app/server/analytics/flares.ts
import { Meal, Symptom } from "@shared/schema";

// Symptoms logged within this many minutes of the first one are one flare
export const FLARE_CLUSTER_MINUTES = 30;

export type Flare = {
  start: Date;
  end: Date;
  symptoms: Symptom[];
  names: string[];
  maxIntensity: number;
};

export function intensityOf(symptom: Symptom): number {
  if (symptom.intensity != null) return symptom.intensity;
  switch (symptom.severity) {
    case "Mild": return 2;
    case "Severe": return 4;
    default: return 3;
  }
}

/** Groups symptoms (any order) into flares. */
export function clusterFlares(symptoms: Symptom[]): Flare[] {
  const sorted = [...symptoms].sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  const flares: Flare[] = [];
  const windowMs = FLARE_CLUSTER_MINUTES * 60 * 1000;
  for (const symptom of sorted) {
    const current = flares[flares.length - 1];
    if (current && symptom.timestamp.getTime() - current.start.getTime() <= windowMs) {
      current.symptoms.push(symptom);
      current.end = symptom.timestamp;
      if (!current.names.includes(symptom.name)) current.names.push(symptom.name);
      current.maxIntensity = Math.max(current.maxIntensity, intensityOf(symptom));
    } else {
      flares.push({ start: symptom.timestamp, end: symptom.timestamp, symptoms: [symptom], names: [symptom.name], maxIntensity: intensityOf(symptom) });
    }
  }
  return flares;
}

export function hoursBetween(earlier: Date, later: Date): number {
  return (later.getTime() - earlier.getTime()) / 3_600_000;
}

/** Meals eaten in the `windowHours` before `at`, most recent first. */
export function mealsBefore(meals: Meal[], at: Date, windowHours: number): Meal[] {
  return meals
    .filter((m) => {
      const h = hoursBetween(m.timestamp, at);
      return h > 0 && h <= windowHours;
    })
    .sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());
}

/** Hours from the most recent prior meal (within the window), or null. */
export function onsetHours(meals: Meal[], at: Date, windowHours: number): number | null {
  const [latest] = mealsBefore(meals, at, windowHours);
  return latest ? hoursBetween(latest.timestamp, at) : null;
}
